File: 111.minimum-depth-of-binary-tree.py
#
# @lc app=leetcode id=111 lang=python3
#
# [111] Minimum Depth of Binary Tree
#

# @lc code=start
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def minDepth(self, root: Optional[TreeNode]) -> int:
        
        if not root:
            return 0
        
        if not root.left and not root.right:
            return 1
        
        left_height = self.minDepth(root.left)
        right_height = self.minDepth(root.right)

        if not root.left:
            return right_height + 1
        if not root.right:
            return left_height + 1
            
        return min(left_height, right_height) + 1
        
# @lc code=end

File: 111.minimum-depth-of-binary-tree.java
/*
 * @lc app=leetcode id=111 lang=java
 *
 * [111] Minimum Depth of Binary Tree
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int minDepth(TreeNode root) {

        if (root == null) {
            return 0;
        } else if (root.left == null && root.right == null) {
            return 1;
        }

        int leftHeight = this.minDepth(root.left);
        int rightHeight = this.minDepth(root.right);

        if (root.left == null) {
            return rightHeight + 1;
        } else if (root.right == null) {
            return leftHeight + 1;
        }

        return Math.min(leftHeight, rightHeight) + 1;
        
    }
}
// @lc code=end

